helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ResourceTopStateAntiAffinityConstraint.java

```diff
@@ -29,25 +29,11 @@
  * score, vice versa.
  */
 class ResourceTopStateAntiAffinityConstraint extends SoftConstraint {
-  private static final float MAX_SCORE = 1f;
-  private static final float MIN_SCORE = 0f;
-
-  ResourceTopStateAntiAffinityConstraint() {
-    super(MAX_SCORE, MIN_SCORE);
-  }
 
   @Override
   protected float getAssignmentScore(AssignableNode node, AssignableReplica replica,
       ClusterContext clusterContext) {
-    if (!replica.isReplicaTopState()) {
-      return (getMaxScore() + getMinScore()) / 2.0f;
-    }
-
-    int curTopPartitionCountForResource = node.getAssignedTopStatePartitionsCount();
-    int doubleMaxTopStateCount = 2 * clusterContext.getEstimatedMaxTopStateCount();
-
-    return Math.max(
-        ((float) doubleMaxTopStateCount - curTopPartitionCountForResource) / doubleMaxTopStateCount,
-        0);
+    // if replica is top state, there're 1 more top state partition hosted on the instance if the assignment happens
```

Comment:
there is?

Also why is it necessarily 1 or 0? I get that if it's not a top state replica, it should be 0. But if it a top state replica, anything above 0 would work? I'm just wondering why it's hard-coded at 1.

Reply:
It's because if the potential assignment pair being evaluated (replica, node) is chosen, it adds "one" more top state partitions + existing top state partitions on the instance.